instance/models/openedx_instance.py

```diff
@@ -249,54 +243,107 @@ def delete(self, *args, **kwargs):
+            ip_address = self.active_appserver.server.public_ip
+        else:
+            backend_name = "be-preliminary-page-{}".format(self.pk)
+            server_name = "preliminary-page"
+            ip_address = settings.PRELIMINARY_PAGE_SERVER_IP
+        if not ip_address:
+            # No active appserver and PRELIMINARY_PAGE_SERVER_IP not set, so there is no backend
+            # we can configure.  Simply return an empty configuration in this case.
+            self.logger.info("PRELIMINARY_PAGE_SERVER_IP is unset, so not configuring a preliminary backend.")
+            return "", ""
+        template = loader.get_template("instance/haproxy/backend.conf")
+        config = template.render(dict(
+            domain=self.domain,
+            backend_name=backend_name,
+            server_name=server_name,
+            ip_address=ip_address,
+        ))
+        domain_names = [
+            self.external_lms_domain, self.external_lms_preview_domain, self.external_studio_domain,
+            self.internal_lms_domain, self.internal_lms_preview_domain, self.internal_studio_domain,
+        ]
+        backend_map = "\n".join(" ".join([domain, backend_name]) for domain in domain_names if domain)
+        return backend_map, config
+
+    @property
+    def appserver_set(self):
+        """
+        Get the set of OpenEdxAppServers owned by this instance.
+        """
+        return self.ref.openedxappserver_set
+
+    def set_appserver_active(self, appserver_id):
+        """
+        Mark the AppServer with the given ID as the active one.
+        """
+        app_server = self.appserver_set.get(pk=appserver_id)  # Make sure the AppServer is owned by this instance
+        self.logger.info('Making %s active for instance %s...', app_server.name, self.name)
         self.active_appserver = app_server
         self.save()
-
+        self.load_balancing_server.reconfigure()
         self.enable_monitoring()
 
+    def set_appserver_inactive(self):
+        """
+        Reset the active AppServer to None.  Called when the active appserver is terminated.
+        """
+        self.disable_monitoring()
+        self.active_appserver = None
+        self.save()
+        self.load_balancing_server.reconfigure()
+
     @log_exception
     def spawn_appserver(self):
         """
         Provision a new AppServer.
 
         Returns the ID of the new AppServer on success or None on failure.
         """
+        if not self.load_balancing_server:
+            self.load_balancing_server = select_load_balancing_server()
+            self.load_balancing_server.reconfigure()
+
+        # We unconditionally set the DNS records here, though this would only be strictly needed
+        # when the first AppServer is spawned.  However, there is no easy way to tell whether the
+        # DNS records have already been successfully set, and it doesn't hurt to alway do it.
+        self.set_dns_records()
```

Comment:
I just realized we don't seem to remove DNS entries from gandi when instances are destroyed. It has nothing to do with this PR, but it would make sense to implement pruning of DNS records when instances are destroyed.